src/native/libs/System.Net.Security.Native.Apple/pal_networkframework.m

```diff
@@ -0,0 +1,433 @@
+        NSNumber* num = nw_framer_options_copy_object_value(framer_options, "GCHANDLE");
+        assert(num != NULL);
+
+        void * ptr;
+        [num getValue:&ptr];
+        size_t size = message_length;
+
+        nw_framer_parse_output(framer, 1, message_length, NULL, ^size_t(uint8_t *buffer, size_t buffer_length, bool is_complete2) {
+            size_t length = buffer_length;
+            (_writeFunc)(ptr, buffer, length);
+            (void)is_complete2;
+            (void)message;
+            return buffer_length;
+        });
+    }
+    else
+    {
+        assert(0);
+    }
+    (int)is_complete;
+};
+
+static nw_framer_stop_handler_t framer_stop_handler = ^bool(nw_framer_t framer) {
+    if (__builtin_available(macOS 12.3, iOS 15.4, tvOS 15.4, watchOS 2.0, *))
+    {
+        size_t gcHandle = 0;
+        nw_protocol_options_t framer_options = nw_framer_copy_options(framer);
+        NSNumber* num = nw_framer_options_copy_object_value(framer_options, "GCHANDLE");
+        if (num != NULL)
+        {
+            [num getValue:&gcHandle];
+            (_statusFunc)(gcHandle, PAL_NwStatusUpdates_FramerStop, 0, 0);
+        }
+    }
+    else
+    {
+        assert(0);
+    }
+
+    return TRUE;
+};
+
+static nw_framer_cleanup_handler_t framer_cleanup_handler = ^(nw_framer_t framer) {
+    (void)framer;
+};
+
+
+// This is called when connection start to set up framer
+static nw_framer_start_handler_t framer_start = ^nw_framer_start_result_t(nw_framer_t framer)
+{
+    assert(_statusFunc != NULL);
+    size_t gcHandle = 0;
+
+    nw_protocol_options_t framer_options = nw_framer_copy_options(framer);
+    NSNumber* num = nw_framer_options_copy_object_value(framer_options, "GCHANDLE");
+    assert(num != NULL);
+
+    [num getValue:&gcHandle];
+
+    // Notify SafeHandle with framer instance so we can submit to it directly.
+    (_statusFunc)(gcHandle, PAL_NwStatusUpdates_FramerStart, (size_t)framer, 0);
+
+    nw_framer_set_output_handler(framer, framer_output_handler);
+
+    nw_framer_set_stop_handler(framer, framer_stop_handler);
+    nw_framer_set_cleanup_handler(framer, framer_cleanup_handler);
+    return nw_framer_start_result_ready;
+};
+
+
+// this takes encrypted input from underlying stream and feeds it to nw_connection.
+PALEXPORT int32_t AppleNetNative_NwProcessInputData(nw_connection_t connection, nw_framer_t framer, const uint8_t * buffer, int bufferLength)
+{
+    if (connection == NULL || framer == NULL)
+    {
+        return -1;
+    }
+
+    nw_framer_message_t message = nw_framer_message_create(framer);
+
+    // There is race condition when connection can fail or be canceled and if it does we fail to create the message here.
+    if (message == NULL)
+    {
+        return -1;
+    }
+
+    uint8_t * copy = NULL;
+    if (bufferLength > 0)
+    {
+        copy = malloc((size_t)bufferLength);
+        memcpy(copy, buffer, bufferLength);
+        nw_framer_message_set_value(message, "DATA", copy,  ^(void* ptr) {
```

Comment:
Are there any situations where this callback won't every be invoked? If that could happen, we'll leak the allocated buffer.